### ISSUE TITLE
Feature request: spatial widget

### ISSUE THREAD

**Issue by florianm** [ckan/ckanext-scheming#11]
Ok this is stretching it a bit, but since scheming tackles custom form widgets, and ckanext-spatial provides this really good to have but (for end users) quite challenging `spatial` field, it would be very useful to integrate both.

A form_snippet spatial.html should allow users to
- paste a valid GeoJSON geometry as per ckanext-spatial, or
- [draw a polygon](http://leafletjs.com/2013/02/20/guest-post-draw.html) on a leaflet map, writing back the geojson geometry of the convex hull (if several polygons are drawn) or
- upload a geojson file with the bounding polygon (or a bunch of features), then save the convex hull's geometry

A display_snippet spatial.html could render the dataset extent just like ckanext-spatial does as a map, hiding the unhelpful WKT blurb. Edit: [done](https://github.com/florianm/ckanext-scheming/tree/11-spatial)
![image](https://cloud.githubusercontent.com/assets/762815/5373181/e309eeb2-808d-11e4-9f73-0abbd81c817a.png)

Also posted to [ckanext-spatial](https://github.com/ckan/ckanext-spatial/issues/91).

**Comment by wardi** [comment]
[here](https://github.com/open-data/ckanext-scheming/issues/11#issuecomment-66448057) I said this should be separate from ckanext-scheming, but this looks so nice I'm changing my mind. It would be great to have a good standard spatial input and view like this in scheming itself, since users are still free to extend or replace it in their own extensions.

If most of this work applies to ckanext-spatial users that might not be interested in scheming, though, then it's probably better not to duplicate it.

**Comment by florianm** [comment]
I agree, I feel this is more a ckanext-scheming feature than a ckanext-spatial one, as it provides a custom widget for a custom field. However, the widget depends on features which are more at home in ckanext-spatial. If that's ok, I'll need to ask you and the ckanext-spatials for advice on the mailing list.

**Comment by florianm** [comment]
Hi all,

could I ask you all, especially the ckanext-spatial and -scheming crew for
advice.

My CKAN users love georeferencing their datasets, but, being environmental
scientists with sometimes limited GIS skills, some struggle with pasting a
correct GeoJSON geometry into the "spatial" extra field. So, with guidance
from Ian I've prototyped a spatial input widget which you can see in action
at [0] and take for a spin at [1] with un/pw "ckan".

The widget still shows the text input for a GeoJSON geometry (enlarged into
a textarea), but also adds a map one can draw the dataset's extent on, and
a select box for often needed, well-defined areas.

This widget depends on a few components which may or may not belong into
ckanext-spatial or -scheming.
ckanext-spatial PR at [4]
- A module spatial_form (based on spatial_query)
- Updated Leaflet.draw, which requires an updated LeafletJS (and breaks
spatial_query)
- Fixing spatial_query to work again with backwards-incompatible
Leaflet.draw

ckanext-scheming feature request at [5]
- A new form snippet spatial.html [3] which depends on spatial_form to
render the map, but also renders a select for pre-defined areas, and
contains the JS to data-bind the inputs together.
Also I've included display snippets, which render the dataset's extent
inline (see video), which replaces the dataset's extent as facet in the
sidebar.

The whole setup works well enough on my end, but before I can send a
clean(er than [4]) PR I'll need your advice on a few points:

Should all components of the spatial widget live in ckanext-scheming, or
which components would be appropriate as PR in ckanext-spatial?

What is the official procedure for upgrading JS dependencies in extensions?
Is there a more graceful way than overwriting files in /vendor/ ?

As for data-binding the three input widgets (map, select, textarea), should
we use heavier artillery like react.js or will we get away with a few
manual JS hooks? Note in my prototype I've only bound map->text and
select->text, but not select (in scheming) to map (in spatial).

As for the select box, that really should also live in the spatial_form JS
module, right? I kept it in ckanext-scheming's form snippet for now, as
it's a ckanext-scheming select widget with data coming out of my custom
data schema definition. Of course that's clunky and hard to maintain, so
I'd rather have a custom vocabulary with an API (for maintenance) and an
autocomplete service (for the input), much like tags, but we'll need three
values (id, label, GeoJSON geometry/string) instead of two, like tags (id
and label).
Would it be feasible to have an autocompleting custom vocabulary for
spatial entities, and how would I implement this without interfering with
models?

Cheers,
Florian


[0] Spatial widget in action (apologies for potato quality)
http://vimeo.com/florianwmayer/ckan-georeferencing
[1] Demo instance http://data-demo.dpaw.wa.gov.au/
[2] spatial_form JS module
https://github.com/florianm/ckanext-spatial/blob/91-spatial-form/ckanext/spatial/public/js/spatial_form.js
[3] spatial form snippet
https://github.com/florianm/ckanext-scheming/blob/master/ckanext/scheming/templates/scheming/form_snippets/spatial.html

[4] PR ckanext-spatial https://github.com/ckan/ckanext-spatial/pull/93/
[5] Feature request at ckanext-scheming
https://github.com/open-data/ckanext-scheming/issues/11

**Comment by MandanaMoshref** [comment]
@florianm Is it possible in your approach to adding marker too? As far as I know, the FeatureCollection and GeometryCollection is not supported through ckanext-spatial (https://github.com/ckan/ckanext-spatial/issues/152#issuecomment-252630806).
In your "spatial_form.js" you aggregate all geometries into multipolygon. 
I would like to have both multipoint and multigeometry but in a way that avoid a mix of both (geometry collection type). So for example, if someone picks a marker automatically leaflet deactivates the BBOX and polygon or vise versa and aggregate corresponding features (multipolygon or multipoints).
Can you advise me on how this can be done in spatial_form.js?
Thanks
Mani

**Comment by florianm** [comment]
@MandanaMoshref thanks for the follow-up! This PR is pretty old and I ran out of bandwidth to bring it to a clean state. We've been using my fork in production since then but are looking to update our setup.

To explain my design choice of using Multipolugons, these simply approximate our datasets best. The raw data are mostly observations with a location (e.g. CSV with lat/lon coordinates and map resource preview), the sites are mostly polygons (e.g. GeoJSON again with a map resource preview, also they show up on my widget), and the dataset extents featured in this PR are the convex hull (polygon) possibly with gaps (multipolygon) around the sites our data come from. Therefore it never made sense to georeference an individual dataset by a point alone. I acknowledge that your use case might be different!

To answer your question, if you could rewrite this PR to use a FeatureCollection instead of a Multipolygon, you may be able to also represent your datasets with a point. You might even get away with leaving all controls enabled.
